File: test/Zorglub.Tests.CSharp/Time/Core/TemporalArithmeticTests.cs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2020 Narvalo.Org. All rights reserved.

namespace Zorglub.Time.Core;

using System.Security.Cryptography;

using static TemporalConstants;

public static partial class TemporalArithmeticTests
{
    // MultiplierData stop at 10.
    private const int RandomFrom = 11;

    public static readonly TheoryData<int> MultiplierData = new()
    { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };

    public static readonly TheoryData<int> ZeroToTwentyThreeData = new()
    { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23 };
}

public partial class TemporalArithmeticTests // TicksPerDay
{
    [Theory, MemberData(nameof(MultiplierData))]
    public static void DivideByTicksPerDay(int mul) =>
        Assert.Equal(mul, TemporalArithmetic.DivideByTicksPerDay(mul * TicksPerDay));

    [Fact]
    public static void DivideByTicksPerDay_Fuzzy()
    {
        // Arrange
        int mul = GetFuzzyMulForTicksPerDay();
        // Act & Assert
        Assert.Equal(mul, TemporalArithmetic.DivideByTicksPerDay(mul * TicksPerDay));
    }

    [Theory, MemberData(nameof(MultiplierData))]
    public static void MultiplyByTicksPerDay(int mul) =>
        Assert.Equal(mul * TicksPerDay, TemporalArithmetic.MultiplyByTicksPerDay(mul));

    [Fact]
    public static void MultiplyByTicksPerDay_Fuzzy()
    {
        // Arrange
        int mul = GetFuzzyMulForTicksPerDay();
        // Act & Assert
        Assert.Equal(mul * TicksPerDay, TemporalArithmetic.MultiplyByTicksPerDay(mul));
    }

    private static int GetFuzzyMulForTicksPerDay()
    {
        const long MaxMul = Int64.MaxValue / TicksPerDay;

        int max = (int)System.Math.Min(Int32.MaxValue, MaxMul);
        return RandomNumberGenerator.GetInt32(RandomFrom, max);
    }
}

public partial class TemporalArithmeticTests // NanosecondsPerHour
{
    [Theory, MemberData(nameof(ZeroToTwentyThreeData))]
    public static void DivideByNanosecondsPerHour(int mul) =>
        Assert.Equal(mul, TemporalArithmetic.DivideByNanosecondsPerHour(mul * NanosecondsPerHour));

    [Theory, MemberData(nameof(ZeroToTwentyThreeData))]
    public static void MultiplyByNanosecondsPerHour(int mul) =>
        Assert.Equal(mul * NanosecondsPerHour, TemporalArithmetic.MultiplyByNanosecondsPerHour(mul));
}

public partial class TemporalArithmeticTests // NanosecondsPerMinute
{
    [Theory, MemberData(nameof(MultiplierData))]
    public static void DivideByNanosecondsPerMinute(int mul) =>
        Assert.Equal(mul, TemporalArithmetic.DivideByNanosecondsPerMinute(mul * NanosecondsPerMinute));

    [Fact]
    public static void DivideByNanosecondsPerMinute_Fuzzy()
    {
        // Arrange
        // In Debug mode, we check that the input is < NanosecondsPerDay,
        // which means that "mul" must be < MinutesPerDay.
        int mul = RandomNumberGenerator.GetInt32(RandomFrom, MinutesPerDay);
        // Act & Assert
        Assert.Equal(mul, TemporalArithmetic.DivideByNanosecondsPerMinute(mul * NanosecondsPerMinute));
    }

    [Theory, MemberData(nameof(MultiplierData))]
    public static void MultiplyByNanosecondsPerMinute(int mul) =>
        Assert.Equal(mul * NanosecondsPerMinute, TemporalArithmetic.MultiplyByNanosecondsPerMinute(mul));

    [Fact]
    public static void MultiplyByNanosecondsPerMinute_Fuzzy()
    {
        // Arrange
        int mul = RandomNumberGenerator.GetInt32(RandomFrom, MinutesPerDay);
        // Act & Assert
        Assert.Equal(mul * NanosecondsPerMinute, TemporalArithmetic.MultiplyByNanosecondsPerMinute(mul));
    }
}

File: test/Zorglub.Tests/Tests/Core/TemporalArithmeticTests.fs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) Tran Ngoc Bich. All rights reserved.

module Zorglub.Tests.Core.TemporalArithmeticTests

open Checked

open System
open System.Security.Cryptography

open Zorglub.Testing

open Zorglub.Time.Core

open Xunit

[<Literal>]
let private RandomFrom = 11
let private maxMul = Int64.MaxValue / TemporalConstants.TicksPerDay

let private getFuzzyMul64ForTicksPerDay () =
    let max = int <| Math.Min(int64(Int32.MaxValue), maxMul)
    int64 <| RandomNumberGenerator.GetInt32(RandomFrom, max)

[<Sealed>]
type MultiplierData() as self =
    inherit TheoryData<int>()
    do
        [for i in 0 .. 10 -> self.Add(i)] |> ignore

[<Sealed>]
type ZeroToTwentyThreeData() as self =
    inherit TheoryData<int>()
    do
        [for i in 0 .. 23 -> self.Add(i)] |> ignore

//
// TicksPerDay
//

[<Theory; ClassData(typeof<MultiplierData>)>]
let ``DivideByTicksPerDay()`` (mul: int) =
    let mul64 = int64(mul)

    TemporalArithmetic.DivideByTicksPerDay(mul64 * TemporalConstants.TicksPerDay) === mul64

[<Fact>]
let ``DivideByTicksPerDay() fuzzy`` () =
    let mul64 = getFuzzyMul64ForTicksPerDay()

    TemporalArithmetic.DivideByTicksPerDay(mul64 * TemporalConstants.TicksPerDay) === mul64

[<Theory; ClassData(typeof<MultiplierData>)>]
let ``MultiplyByTicksPerDay()`` (mul: int) =
    let mul64 = int64(mul)

    TemporalArithmetic.MultiplyByTicksPerDay(mul64) === mul64 * TemporalConstants.TicksPerDay

[<Fact>]
let ``MultiplyByTicksPerDay() fuzzy`` () =
    let mul64 = getFuzzyMul64ForTicksPerDay()

    TemporalArithmetic.MultiplyByTicksPerDay(mul64) === mul64 * TemporalConstants.TicksPerDay

//
// NanosecondsPerHour
//

[<Theory; ClassData(typeof<ZeroToTwentyThreeData>)>]
let ``DivideByNanosecondsPerHour()`` (mul: int) =
    TemporalArithmetic.DivideByNanosecondsPerHour(int64(mul) * TemporalConstants.NanosecondsPerHour) === mul

[<Theory; ClassData(typeof<ZeroToTwentyThreeData>)>]
let ``MultiplyByNanosecondsPerHour()`` (mul: int) =
    TemporalArithmetic.MultiplyByNanosecondsPerHour(mul) === int64(mul) * TemporalConstants.NanosecondsPerHour

//
// NanosecondsPerMinute
//

[<Theory; ClassData(typeof<MultiplierData>)>]
let ``DivideByNanosecondsPerMinute()`` (mul: int) =
    TemporalArithmetic.DivideByNanosecondsPerMinute(int64(mul) * TemporalConstants.NanosecondsPerMinute) === mul

[<Fact>]
let ``DivideByNanosecondsPerMinute() fuzzy`` () =
    // In Debug mode, we check that the input is < NanosecondsPerDay,
    // which means that "mul" must be < MinutesPerDay.
    let mul = RandomNumberGenerator.GetInt32(RandomFrom, TemporalConstants.MinutesPerDay)

    TemporalArithmetic.DivideByNanosecondsPerMinute(int64(mul) * TemporalConstants.NanosecondsPerMinute) === mul

[<Theory; ClassData(typeof<MultiplierData>)>]
let ``MultiplyByNanosecondsPerMinute()`` (mul: int) =
    TemporalArithmetic.MultiplyByNanosecondsPerMinute(mul) === int64(mul) * TemporalConstants.NanosecondsPerMinute

[<Fact>]
let ``MultiplyByNanosecondsPerMinute() fuzzy`` () =
    let mul = RandomNumberGenerator.GetInt32(RandomFrom, TemporalConstants.MinutesPerDay)

    TemporalArithmetic.MultiplyByNanosecondsPerMinute(mul) === int64(mul) * TemporalConstants.NanosecondsPerMinute
